### ISSUE TITLE
No generalization for recursive functions

### ISSUE THREAD

**Issue by smimram** [savonet/liquidsoap#921]
Recursive functions do not seem to have their type correctly generalized. Consider
```
def f(x) = x end
def rec g(x) = x end
```
we respectively get the types
```
f     : ('a) -> 'a
g     : (?A) -> ?A
```
which means that `g` is not polymorphic...

**Comment by smimram** [comment]
The following fixes the bug but I am not sure about the implications:
```
diff --git a/src/lang/lang_values.ml b/src/lang/lang_values.ml
index 9c81d7ee..ee8eb101 100644
--- a/src/lang/lang_values.ml
+++ b/src/lang/lang_values.ml
@@ -546,6 +546,7 @@ let (>:) = T.(>:)
 let rec value_restriction t = match t.term with
   | Var _ -> true
   | Fun _ -> true
+  | RFun _ -> true
   | List l | Tuple l -> List.for_all value_restriction l
   | _ -> false
 
```
@dbaelde could you comment on it?

**Comment by smimram** [comment]
With the above patch we seemingly get the right type but the following is not working
```
def rec recid(x)
  x
end
ignore(recid(3) == 3)
ignore(recid("a") == "a")
```
although `recid` has type
```
('a) -> 'a
```

**Comment by smimram** [comment]
Narrowing it down. The following more minimal example is working
```
def id (x) = x end
_ = id(3)
_ = id("a")
```
but not
```
def rec recid (x) = x end
_ = recid(3)
_ = recid("a")
```
However, the following is working
```
def rec recid (x) = 5 end
_ = recid(3)
_ = recid("a")
```
Somehow the generalization of the type does not refresh the type of the variable `x` in the body of `recid`. This is validated with
```
export LIQUIDSOAP_DEBUG_LANG=true
```
with the following patch to show runtime instantiations:
```
diff --git a/src/lang/lang_values.ml b/src/lang/lang_values.ml
index ee8eb101..e6db148a 100644
--- a/src/lang/lang_values.ml
+++ b/src/lang/lang_values.ml
@@ -863,6 +863,9 @@ let instantiate ~generalized def =
 let lookup env var ty =
   let generalized,def = List.assoc var env in
   let v = instantiate ~generalized def in
+    if debug then
+      Printf.eprintf "Runtime instantiate %s[%d] : %s becomes %s\n"
+        var (T.deref def.V.t).T.level (T.print def.V.t) (T.print v.V.t);
     if debug then
       Printf.eprintf
         "Runtime instantiation of %s: %s targets %s.\n"
```

**Comment by dbaelde** [comment]
I have looked at it but I do not understand some code relating to recursive functions, which was actually written by @toots in 8862d523ac32e1fb9f9b36499153fc7cc9bdfa40. In particular, I would like to understand why `RFun` takes a `unit->term` for the function body, rather than just a `term` as in `Fun`.

**Comment by dbaelde** [comment]
Actually, looking at the log I see that the recursive function is actually not generalized. This can also be seen directly seen with `liquidsoap -t 'def rec id(x) = x end'`.

**Comment by smimram** [comment]
It is if you apply the above patch!

**Comment by dbaelde** [comment]
Sorry Sam, I just realized that. Having applied the patch I now see the problem that you have: it is not a typing but a runtime issue. I will look more into it.

**Comment by smimram** [comment]
Great, thanks and welcome back! ;)

**Comment by dbaelde** [comment]
I have difficulties wrapping my head around the runtime type propagation, and I have to stop for now without a solution. I cannot see how the variables inside the non-recursive function bodies receive their instantiated types when evaluating the function: the same should happen with recursive functions, but it does not (instead the original type variable is found, and as a result it becomes bound to the first instantiation, creating the error when trying to bind it to the second instantiation).

I still suspect that the problem is with the `unit->term` body of recursive functions: I do not see why it is so; I find it suspicious that every instance of the body-function has a new kind (generated by `mk` inside the `fn` function of `mk_rec_fun`; another probable bug is the empty `gen` argument of the `Let` inside that `fn` function, though it should not be the issue here since we do not have actual recursive calls.

**Comment by smimram** [comment]
Ok, so I tried to remove `unit->term`, see #934 and here is the problem. At execution time, we have to instantiate type classes, which requires going through terms replacing types, which makes the naive implementation loop in `V.map_types` called by `instantiate`. For now, I have changed `map_types` so that it does not recurse in closures ([see there](https://github.com/savonet/liquidsoap/pull/934/files#diff-84ccbca8728deab9d31a31fe3cbdc986R493)), but this is not a clean solution... I guess this is the reason for the `unit -> term`.

Anybody with a better idea?
